In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 4: Security Through Obscurity ---</h2><p>Finally, you come across an information kiosk with a list of rooms.  Of course, the list is encrypted and full of decoy data, but the instructions to decode the list are barely hidden nearby.  Better remove the decoy data first.</p>
<p>Each room consists of an encrypted name (lowercase letters separated by dashes) followed by a dash, a sector ID, and a checksum in square brackets.</p>
<p>A room is real (not a decoy) if the checksum is the five most common letters in the encrypted name, in order, with ties broken by alphabetization.  For example:</p>
<ul>
<li><code>aaaaa-bbb-z-y-x-123[abxyz]</code> is a real room because the most common letters are <code>a</code> (5), <code>b</code> (3), and then a tie between <code>x</code>, <code>y</code>, and <code>z</code>, which are listed alphabetically.</li>
<li><code>a-b-c-d-e-f-g-h-987[abcde]</code> is a real room because although the letters are all tied (1 of each), the first five are listed alphabetically.</li>
<li><code>not-a-real-room-404[oarel]</code> is a real room.</li>
<li><code>totally-real-room-200[decoy]</code> is not.</li>
</ul>
<p>Of the real rooms from the list above, the sum of their sector IDs is <code>1514</code>.</p>
<p>What is the <em>sum of the sector IDs of the real rooms</em>?</p>
</article>


In [30]:
from collections import Counter


def is_real_room(room: str) -> bool:
    text, cs = re.split(r"\d+", room.replace("-", ""))
    cs = cs[1:-1]

    count = Counter(text)
    prev = inf
    for l in cs:
        if l not in count or count[l] > prev:
            return False
        prev = count.pop(l)
    return all(v <= prev for v in count.values())


assert is_real_room("aaaaa-bbb-z-y-x-123[abxyz]")
assert is_real_room("a-b-c-d-e-f-g-h-987[abcde]")
assert is_real_room("not-a-real-room-404[oarel]")
assert not is_real_room("totally-real-room-200[decoy]")

In [40]:
digest = """
aaaaa-bbb-z-y-x-123[abxyz]
a-b-c-d-e-f-g-h-987[abcde]
not-a-real-room-404[oarel]
totally-real-room-200[decoy]
"""


def sum_of_sector_ids_real_rooms(digest: str) -> int:
    return sum(
        int(re.findall(r"\d+", line)[0])
        for line in digest.strip().splitlines()
        if is_real_room(line)
    )


assert sum_of_sector_ids_real_rooms(digest) == 1514

In [43]:
with open("../input/day4.txt") as f:
    digest_f = f.read()

sum_of_sector_ids_real_rooms(digest_f)

158835

<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>158835</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>With all the decoy data out of the way, it's time to decrypt this list and get moving.</p>
<p>The room names are encrypted by a state-of-the-art <a href="https://en.wikipedia.org/wiki/Caesar_cipher">shift cipher</a>, which is nearly unbreakable without <span title="...or, like, half of a napkin.">the right software</span>. However, the information kiosk designers at Easter Bunny HQ were not expecting to deal with a master cryptographer like yourself.</p>
<p>To decrypt a room name, rotate each letter forward through the alphabet a number of times equal to the room's sector ID.  <code>A</code> becomes <code>B</code>, <code>B</code> becomes <code>C</code>, <code>Z</code> becomes <code>A</code>, and so on. Dashes become spaces.</p>
<p>For example, the real name for <code>qzmt-zixmtkozy-ivhz-343</code> is <code>very encrypted name</code>.</p>
<p><em>What is the sector ID</em> of the room where North Pole objects are stored?</p>
</article>

</main>


In [58]:
e = "qzmt-zixmtkozy-ivhz-343"
d = "very encrypted name"

from string import ascii_lowercase as lw


def decrypt(e: str) -> str:
    n = len(lw)
    sid = int(re.findall(r"\d+", e)[0])
    return " ".join(
        "".join(lw[(lw.find(l) + sid) % n] for l in w) for w in e.split("-")[:-1]
    )


decrypt(e)

'very encrypted name'

In [67]:
for line in digest_f.strip().splitlines():
    text = decrypt(line)
    if "northpole" in text:
        print(f"{text=}")
        print(f"{line=}")
        sid = int(re.findall(r"\d+", line)[0])
        print(f"{sid=}")

text='northpole object storage'
line='ijmockjgz-jwezxo-nojmvbz-993[jozmb]'
sid=993


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>993</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2016">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="4/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Security+Through+Obscurity%22+%2D+Day+4+%2D+Advent+of+Code+2016&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2016%2Fday%2F4&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Security+Through+Obscurity%22+%2D+Day+4+%2D+Advent+of+Code+2016+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2016%2Fday%2F4';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
